src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -79,29 +267,50 @@ extern "C" SortHandle* GetSortHandle(const char* lpLocaleName)
 extern "C" void CloseSortHandle(SortHandle* pSortHandle)
 {
     ucol_close(pSortHandle->regular);
-    ucol_close(pSortHandle->ignoreCase);
-
     pSortHandle->regular = nullptr;
-    pSortHandle->ignoreCase = nullptr;
+
+    TCollatorMap::iterator it;
+    for (it = pSortHandle->collatorsPerOption.begin(); it != pSortHandle->collatorsPerOption.end(); it++)
+    {
+        ucol_close(it->second);
+    }
 
     delete pSortHandle;
 }
 
-const UCollator* GetCollatorFromSortHandle(const SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
+const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    if ((options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase)
+    UCollator* pCollator;
+    if (options == 0)
+    {
+        pCollator = pSortHandle->regular;
+    }
+    else
     {
-        return pSortHandle->ignoreCase;
+        pthread_mutex_lock(&pSortHandle->collatorsLockObject);
+
+        TCollatorMap::iterator entry = pSortHandle->collatorsPerOption.find(options);
+        if (entry == pSortHandle->collatorsPerOption.end())
+        {
+            pCollator = CloneCollatorWithOptions(pSortHandle->regular, options, pErr);
```

Comment:
This is now allocating a new collator where it wasn't before, right?  Are we cleaning it up appropriately at all call sites?  And does this have any measurable performance impact?